app/js/models/addon_generator.js

```diff
@@ -0,0 +1,225 @@
+/*global Model*/
+
+/*global JSZip*/
+
+export default class AddonGenerator extends Model {
+  constructor(properties) {
+    super(properties);
+
+    this.operations = [];
+
+    this.id = this.id || ('addon' + Math.round(Math.random() * 100000000));
+    this.name = this.name || this.id;
+
+    this.packageMetadata = {
+      installOrigin: 'http://gaiamobile.org',
+      manifestURL: 'app://' + this.id + '.gaiamobile.org/update.webapp',
+      version: 1
+    };
+
+    this.packageManifest = {
+      name: this.name,
+      package_path: '/application.zip'
+    };
+
+    this.manifest = {
+      name: this.name,
+      description: 'Generated by Customizer',
+      role: 'addon',
+      type: 'certified',
+      origin: 'app://' + this.id + '.gaiamobile.org',
+      customizations: [{
+        filter: window.location.origin,
+        scripts: ['main.js']
+      }]
+    };
+  }
+
+  generate() {
+    var script =
+`/*=AddonGenerator*/
+var selector = '${this.getSelector()}';
+var el = document.querySelector(selector);
+var mo = new MutationObserver(function() {
+  var newEl = document.querySelector(selector);
+  if (newEl !== el) {
+    el = newEl;
+    setTimeout(exec, 1);
+  }
+});
+mo.observe(document.documentElement, {
+  childList: true,
+  attributes: true,
+  characterData: true,
+  subtree: true
+});
+exec();
+function exec() {
+${this.operations.join('\n')}
+}
+/*==*/`;
+
+    console.log('******** Generated SCRIPT ********', script);
+
+    var applicationZip = new JSZip();
+    applicationZip.file('manifest.webapp', JSON.stringify(this.manifest));
+    applicationZip.file('main.js', script);
+
+    var packageZip = new JSZip();
+    packageZip.file('metadata.json', JSON.stringify(this.packageMetadata));
+    packageZip.file('update.webapp', JSON.stringify(this.packageManifest));
+    packageZip.file('application.zip', applicationZip.generate({ type: 'arraybuffer' }));
+
+    return packageZip.generate({ type: 'arraybuffer' });
+  }
+
+  getSelector() {
+    return getSelector(this.target);
+  }
+
+  addEventListener(eventName, callback) {
+    this.operations.push(
+`/*=AddonGenerator::addEventListener*/
+el.addEventListener('${eventName}', ${callback});
+/*==*/`
+    );
+  }
+
+  appendChild(childNodeName) {
```

Comment:
You might want to rename these to something like ```op***```, eg. ```opAppendChild```. From outside code, you may mistake them for actual DOM operations, rather than just string appending.